File: stride-agent/app/finance.py
# Copyright 2026 Google LLC
# Licensed under the Apache License, Version 2.0 (the "License");
import math

PER_MONTH = {
    'Weekly': 52 / 12,
    'Biweekly': 26 / 12,
    'Semi-monthly': 2.0,
    'Monthly': 1.0,
    'Quarterly': 1 / 3,
    'Annual': 1 / 12,
    'One-time': 0.0,
}

def amort(principal: float, apr_pct: float, months: int) -> float:
    principal = max(0.0, principal)
    months = max(0, int(round(months)))
    if months <= 0:
        return 0.0
    r = apr_pct / 100.0 / 12.0
    if r == 0:
        return principal / months
    try:
        return (principal * r) / (1.0 - math.pow(1.0 + r, -months))
    except (ZeroDivisionError, OverflowError):
        return principal / months

def monthly_equiv(frequency: str, amount: float) -> float:
    return amount * PER_MONTH.get(frequency, 1.0)

def income_at(month_idx: int, streams: list) -> float:
    total = 0.0
    for s in streams:
        if not s.get('active', True):
            continue
        start = int(round(s.get('start', 0)))
        duration = s.get('duration')
        
        # check if active in this month
        if duration is None:
            within = month_idx >= start
        else:
            within = start <= month_idx < start + int(round(duration))
            
        if not within:
            continue
            
        freq = s.get('frequency', 'Monthly')
        amt = float(s.get('amount', 0.0))
        if freq == 'One-time':
            if month_idx == start:
                total += amt
        else:
            total += monthly_equiv(freq, amt)
    return total

def debt_outflow(debts: list) -> float:
    total = 0.0
    for d in debts:
        if not d.get('active', True):
            continue
        total += float(d.get('minPayment', 0.0))
    return total

def milestone_events(g: dict, streams: list, horizon: int) -> dict:
    one_times = []
    recurrings = []
    
    kind = g.get('kind')
    if kind != 'template':
        # Simple Goal
        one_times.append({'m': int(round(g.get('month', 0))), 'amt': float(g.get('amount', 0.0))})
        return {'one_times': one_times, 'recurrings': recurrings}
        
    # Template Goal
    template = g.get('template')
    params = g.get('params', {})
    m = int(round(params.get('month', 0)))
    
    if template == 'buy_car':
        price = float(params.get('price', 0.0))
        down = float(params.get('down', 0.0))
        trade_in = float(params.get('tradeIn', 0.0))
        apr = float(params.get('apr', 0.0))
        term = int(params.get('term', 1))
        insurance = float(params.get('insurance', 0.0))
        gas = float(params.get('gas', 0.0))
        
        principal = max(0.0, price - down - trade_in)
        pay = amort(principal, apr, term)
        
        one_times.append({'m': m, 'amt': down})
        if trade_in > 0:
            one_times.append({'m': m, 'amt': -trade_in})
            
        recurrings.append({'from': m + 1, 'to': m + 1 + term, 'amt': pay})
        recurrings.append({'from': m, 'to': horizon, 'amt': insurance})
        recurrings.append({'from': m, 'to': horizon, 'amt': gas})
        
    elif template == 'buy_house':
        price = float(params.get('price', 0.0))
        down = float(params.get('down', 0.0))
        apr = float(params.get('apr', 0.0))
        amort_years = int(params.get('amortYears', 30))
        closing_pct = float(params.get('closingPct', 0.0))
        tax_pct = float(params.get('taxPct', 0.0))
        insurance = float(params.get('insurance', 0.0))
        maint_pct = float(params.get('maintPct', 0.0))
        hoa = float(params.get('hoa', 0.0))
        
        months = amort_years * 12
        principal = max(0.0, price - down)
        pay = amort(principal, apr, months)
        
        one_times.append({'m': m, 'amt': down})
        one_times.append({'m': m, 'amt': (price * closing_pct) / 100.0})
        
        recurrings.append({'from': m + 1, 'to': m + 1 + months, 'amt': pay})
        recurrings.append({'from': m, 'to': horizon, 'amt': (price * tax_pct) / 100.0 / 12.0})
        recurrings.append({'from': m, 'to': horizon, 'amt': insurance})
        recurrings.append({'from': m, 'to': horizon, 'amt': (price * maint_pct) / 100.0 / 12.0})
        if hoa > 0:
            recurrings.append({'from': m, 'to': horizon, 'amt': hoa})
            
    elif template == 'sabbatical':
        duration = int(params.get('duration', 6))
        pause_key = params.get('pauseKey')
        travel = float(params.get('travel', 0.0))
        healthcare = float(params.get('healthcare', 0.0))
        
        # Find stream to pause
        paused = 0.0
        if pause_key:
            for s in streams:
                if s.get('key') == pause_key:
                    paused = monthly_equiv(s.get('frequency', 'Monthly'), float(s.get('amount', 0.0)))
                    break
        
        recurrings.append({'from': m, 'to': m + duration, 'amt': paused})
        recurrings.append({'from': m, 'to': m + duration, 'amt': travel})
        recurrings.append({'from': m, 'to': m + duration, 'amt': healthcare})
        
    elif template == 'wedding':
        price = float(params.get('price', 0.0))
        contribution = float(params.get('contribution', 0.0))
        one_times.append({'m': m, 'amt': price})
        if contribution > 0:
            one_times.append({'m': m, 'amt': -contribution})
            
    elif template == 'savings_goal':
        monthly_contrib = float(params.get('monthlyContribution', 0.0))
        initial_dep = float(params.get('initialDeposit', 0.0))
        if monthly_contrib > 0:
            recurrings.append({'from': 0, 'to': m, 'amt': -monthly_contrib})
        if initial_dep > 0:
            one_times.append({'m': 0, 'amt': -initial_dep})
            
    elif template == 'custom':
        prims = params.get('prims', [])
        for pr in prims:
            kind = pr.get('kind')
            pm = int(round(pr.get('month', 0)))
            if kind == 'once':
                one_times.append({'m': pm, 'amt': float(pr.get('amt', 0.0))})
            elif kind == 'recur':
                recurrings.append({'from': pm, 'to': pm + int(round(pr.get('months', 1))), 'amt': float(pr.get('amt', 0.0))})
            elif kind == 'loan':
                principal = float(pr.get('principal', 0.0))
                apr = float(pr.get('apr', 0.0))
                term = int(pr.get('term', 1))
                pay = amort(principal, apr, term)
                recurrings.append({'from': pm + 1, 'to': pm + 1 + term, 'amt': pay})
                
    return {'one_times': one_times, 'recurrings': recurrings}

def build_series(
    streams: list,
    spending: float,
    debts: list,
    goals: list,
    statuses: list,
    horizon: int,
    start_bal: float
) -> list:
    arr = []
    bal = start_bal
    debt_out = debt_outflow(debts)
    active_goals = [g for g in goals if g.get('status') in statuses]
    evs = [milestone_events(g, streams, horizon) for g in active_goals]
    
    for i in range(horizon):
        if i > 0:
            bal += income_at(i, streams) - spending - debt_out
        for ev in evs:
            for o in ev['one_times']:
                if o['m'] == i:
                    bal -= o['amt']
            for r in ev['recurrings']:
                if r['from'] <= i < r['to']:
                    bal -= r['amt']
        arr.append(bal)
    return arr

def simulate_debts(debts: list, extra: float, strategy: str) -> dict:
    bals = []
    for d in debts:
        if d.get('active', True) and float(d.get('balance', 0.0)) > 0:
            bals.append({
                'apr': float(d.get('apr', 0.0)),
                'min': float(d.get('minPayment', 0.0)),
                'bal': float(d.get('balance', 0.0))
            })
    
    series = []
    total_interest = 0.0
    month = 0
    max_months = 600
    
    start_total = sum(d['bal'] for d in bals)
    series.append(start_total)
    
    while any(d['bal'] > 0.5 for d in bals) and month < max_months:
        month += 1
        budget = sum(d['min'] for d in bals if d['bal'] > 0.5) + extra
        
        for d in bals:
            if d['bal'] <= 0.5:
                continue
            interest = (d['bal'] * d['apr']) / 100.0 / 12.0
            d['bal'] += interest
            total_interest += interest
            
        # sort order of payments
        active_debts = [d for d in bals if d['bal'] > 0.5]
        if strategy == 'snowball':
            # sort by balance ascending
            active_debts.sort(key=lambda x: x['bal'])
        else:
            # avalanche: sort by APR descending
            active_debts.sort(key=lambda x: x['apr'], reverse=True)
            
        for d in active_debts:
            if budget <= 0:
                break
            pay = min(budget, d['bal'])
            d['bal'] -= pay
            budget -= pay
            
        series.append(sum(max(0.0, d['bal']) for d in bals))
        
    return {
        'months': None if month >= max_months else month,
        'totalInterest': total_interest,
        'series': series
    }

def start_balance(accounts: list, fallback: float = 8200.0) -> float:
    if not accounts:
        return fallback
    return sum(float(x.get('balance', 0.0)) for x in accounts)

File: src/lib/finance.ts
import type {
  Account,
  Debt,
  Goal,
  GoalStatus,
  IncomeStream,
  MilestoneEvents,
} from '../types';
import { monthLabel, money } from './format';

export const PER_MONTH: Record<string, number> = {
  Weekly: 52 / 12,
  Biweekly: 26 / 12,
  'Semi-monthly': 2,
  Monthly: 1,
  Quarterly: 1 / 3,
  Annual: 1 / 12,
  'One-time': 0,
};

// Standard loan amortization payment for a fixed-rate, fixed-term loan.
export function amort(principal: number, aprPct: number, months: number): number {
  principal = Math.max(0, principal);
  months = Math.round(months);
  if (months <= 0) return 0;
  const r = aprPct / 100 / 12;
  if (r === 0) return principal / months;
  return (principal * r) / (1 - Math.pow(1 + r, -months));
}

export function monthlyEquiv(stream: IncomeStream): number {
  return (+stream.amount || 0) * (PER_MONTH[stream.frequency] ?? 1);
}

export function incomeAt(i: number, streams: IncomeStream[]): number {
  let sum = 0;
  for (const s of streams) {
    if (s.active === false) continue;
    const st = Math.round(s.start);
    const within = i >= st && (s.duration == null || i < st + Math.round(s.duration));
    if (!within) continue;
    if (s.frequency === 'One-time') {
      if (i === st) sum += +s.amount || 0;
    } else sum += monthlyEquiv(s);
  }
  return sum;
}

export function steadyIncome(streams: IncomeStream[]): number {
  return incomeAt(0, streams);
}

export function debtOutflow(debts: Debt[]): number {
  let sum = 0;
  for (const d of debts) {
    if (d.active === false) continue;
    sum += +d.minPayment || 0;
  }
  return sum;
}

export function debtInterest(d: Debt): number {
  return ((+d.balance || 0) * (+d.apr || 0)) / 100 / 12;
}

export function isDebtGrowing(d: Debt): boolean {
  return debtInterest(d) > (+d.minPayment || 0) + 0.5;
}

export function breakEvenPayment(d: Debt): number {
  return Math.ceil(debtInterest(d) / 5) * 5;
}

export function startBalance(accounts: Account[], fallback = 8200): number {
  if (!accounts || !accounts.length) return fallback;
  return accounts.reduce((s, x) => s + (+x.balance || 0), 0);
}

export function incomeTimeframe(s: IncomeStream): string {
  const start = Math.round(s.start);
  const freq = s.frequency && s.frequency !== 'Monthly' ? s.frequency : null;
  let base: string;
  if (s.duration == null) {
    base = start === 0 ? 'Ongoing' : 'From ' + monthLabel(start);
  } else {
    const dur = Math.round(s.duration);
    base = dur + (dur === 1 ? ' mo' : ' mos') + ' · ' + monthLabel(start);
  }
  return freq ? freq + ' · ' + base : base;
}

export function mMonth(g: Goal): number {
  if (g.kind !== 'template') return Math.round(g.month ?? 0);
  if (g.template === 'custom') {
    const prims = g.params?.prims || [];
    if (prims.length === 0) return 0;
    return Math.min(...prims.map((pr) => Math.round(pr.month || 0)));
  }
  return Math.round(((g.params?.month as number) ?? 0));
}

export function mFirstWord(g: Goal): string {
  const w = (g.name || '').split(' ')[0];
  return w === 'New' ? g.name.split(' ')[1] || 'Goal' : w;
}

// Net cash outflow events for a milestone (amt > 0 reduces balance).
export function milestoneEvents(g: Goal, streams: IncomeStream[], horizon: number): MilestoneEvents {
  const oneTimes: MilestoneEvents['oneTimes'] = [];
  const recurrings: MilestoneEvents['recurrings'] = [];
  if (g.kind !== 'template') {
    oneTimes.push({ m: Math.round(g.month), amt: g.amount });
    return { oneTimes, recurrings };
  }
  const p = g.params || {};
  const m = Math.round((p.month as number) || 0);
  if (g.template === 'buy_car') {
    const principal = Math.max(0, (+(p.price as number) || 0) - (+(p.down as number) || 0) - (+(p.tradeIn as number) || 0));
    const pay = amort(principal, +(p.apr as number) || 0, +(p.term as number) || 1);
    oneTimes.push({ m, amt: +(p.down as number) || 0 });
    if (+(p.tradeIn as number)) oneTimes.push({ m, amt: -(+(p.tradeIn as number)) });
    recurrings.push({ from: m + 1, to: m + 1 + Math.round(+(p.term as number) || 0), amt: pay });
    recurrings.push({ from: m, to: horizon, amt: +(p.insurance as number) || 0 });
    recurrings.push({ from: m, to: horizon, amt: +(p.gas as number) || 0 });
  } else if (g.template === 'buy_house') {
    const months = Math.round((+(p.amortYears as number) || 30) * 12);
    const principal = Math.max(0, (+(p.price as number) || 0) - (+(p.down as number) || 0));
    const pay = amort(principal, +(p.apr as number) || 0, months);
    oneTimes.push({ m, amt: +(p.down as number) || 0 });
    oneTimes.push({ m, amt: ((+(p.price as number) || 0) * (+(p.closingPct as number) || 0)) / 100 });
    recurrings.push({ from: m + 1, to: m + 1 + months, amt: pay });
    recurrings.push({ from: m, to: horizon, amt: ((+(p.price as number) || 0) * (+(p.taxPct as number) || 0)) / 100 / 12 });
    recurrings.push({ from: m, to: horizon, amt: +(p.insurance as number) || 0 });
    recurrings.push({ from: m, to: horizon, amt: ((+(p.price as number) || 0) * (+(p.maintPct as number) || 0)) / 100 / 12 });
    if (+(p.hoa as number)) recurrings.push({ from: m, to: horizon, amt: +(p.hoa as number) });
  } else if (g.template === 'sabbatical') {
    const dur = Math.round(+(p.duration as number) || 6);
    const src = (streams || []).find((x) => x.key === p.pauseKey);
    const paused = src ? monthlyEquiv(src) : 0;
    recurrings.push({ from: m, to: m + dur, amt: paused });
    recurrings.push({ from: m, to: m + dur, amt: +(p.travel as number) || 0 });
    recurrings.push({ from: m, to: m + dur, amt: +(p.healthcare as number) || 0 });
  } else if (g.template === 'wedding') {
    oneTimes.push({ m, amt: +(p.price as number) || 0 });
    if (+(p.contribution as number)) {
      oneTimes.push({ m, amt: -(+(p.contribution as number)) });
    }
  } else if (g.template === 'savings_goal') {
    // No target-date deduction (since it is a savings bucket, not a payment/purchase)
    // Instead, support recurring/lump-sum positive contributions (negative outflows)
    if (+(p.monthlyContribution as number)) {
      recurrings.push({ from: 0, to: m, amt: -(+(p.monthlyContribution as number)) });
    }
    if (+(p.initialDeposit as number)) {
      oneTimes.push({ m: 0, amt: -(+(p.initialDeposit as number)) });
    }
  } else if (g.template === 'custom') {
    for (const pr of p.prims || []) {
      const pm = Math.round(pr.month || 0);
      if (pr.kind === 'once') oneTimes.push({ m: pm, amt: pr.amt || 0 });
      else if (pr.kind === 'recur') recurrings.push({ from: pm, to: pm + Math.round(pr.months || 1), amt: pr.amt || 0 });
      else if (pr.kind === 'loan') {
        const pay = amort(pr.principal || 0, pr.apr || 0, pr.term || 1);
        recurrings.push({ from: pm + 1, to: pm + 1 + Math.round(pr.term || 0), amt: pay });
      }
    }
  }
  return { oneTimes, recurrings };
}

export interface MilestoneSummary {
  oneTime: number;
  monthlyNow: number;
  monthlyPeak: number;
}

export function milestoneSummary(g: Goal, streams: IncomeStream[], horizon: number): MilestoneSummary {
  const ev = milestoneEvents(g, streams, horizon);
  let oneTime = 0;
  for (const o of ev.oneTimes) oneTime += o.amt;
  let monthlyNow = 0,
    monthlyPeak = 0;
  const startMonth = mMonth(g);
  for (let i = startMonth; i < horizon; i++) {
    let mo = 0;
    for (const r of ev.recurrings) if (i >= r.from && i < r.to) mo += r.amt;
    if (i === startMonth) monthlyNow = mo;
    monthlyPeak = Math.max(monthlyPeak, mo);
  }
  return { oneTime, monthlyNow, monthlyPeak };
}

export function buildSeries(
  streams: IncomeStream[],
  spending: number,
  debts: Debt[],
  goals: Goal[],
  statuses: GoalStatus[],
  horizon: number,
  startBal: number,
  onlyGoalKey?: string,
  excludeGoalKey?: string
): number[] {
  const arr: number[] = [];
  let bal = startBal;
  const debtOut = debtOutflow(debts);
  const active = goals.filter((g) => {
    if (excludeGoalKey && g.key === excludeGoalKey) return false;
    if (onlyGoalKey) {
      return g.status === 'committed' || g.key === onlyGoalKey;
    }
    return statuses.indexOf(g.status) >= 0;
  });
  const evs = active.map((g) => milestoneEvents(g, streams, horizon));
  for (let i = 0; i < horizon; i++) {
    if (i > 0) bal += incomeAt(i, streams) - spending - debtOut;
    for (const ev of evs) {
      for (const o of ev.oneTimes) if (o.m === i) bal -= o.amt;
      for (const r of ev.recurrings) if (i >= r.from && i < r.to) bal -= r.amt;
    }
    arr.push(bal);
  }
  return arr;
}

export interface DebtSimResult {
  months: number | null;
  totalInterest: number;
  series: number[];
}

// Simulate paying down debts under a strategy with an extra monthly budget.
export function simulateDebts(debts: Debt[], extra: number, strategy: 'avalanche' | 'snowball'): DebtSimResult {
  let bals = debts
    .filter((d) => d.active !== false && (+d.balance || 0) > 0)
    .map((d) => ({ apr: +d.apr || 0, min: +d.minPayment || 0, bal: +d.balance || 0 }));
  const series: number[] = [];
  let totalInterest = 0;
  let month = 0;
  const MAX = 600;
  const start = bals.reduce((s, d) => s + d.bal, 0);
  series.push(start);
  while (bals.some((d) => d.bal > 0.5) && month < MAX) {
    month++;
    let budget = bals.reduce((s, d) => s + (d.bal > 0.5 ? d.min : 0), 0) + extra;
    for (const d of bals) {
      if (d.bal <= 0.5) continue;
      const interest = (d.bal * d.apr) / 100 / 12;
      d.bal += interest;
      totalInterest += interest;
    }
    let order = bals.filter((d) => d.bal > 0.5).slice();
    order.sort((a, b) => (strategy === 'snowball' ? a.bal - b.bal : b.apr - a.apr));
    for (const d of order) {
      if (budget <= 0) break;
      const pay = Math.min(budget, d.bal);
      d.bal -= pay;
      budget -= pay;
    }
    series.push(bals.reduce((s, d) => s + Math.max(0, d.bal), 0));
  }
  return { months: month >= MAX ? null : month, totalInterest, series };
}

export interface GoalMetrics {
  monthlyToward: number;
  remaining: number;
  monthsNeeded: number;
  ahead: number;
  tone: string;
  pillBg: string;
  statusLabel: string;
  hitDate: string;
  pct: number;
  sub: string;
}

export function metrics(g: { amount: number; saved: number; ratio: number; month: number }, net: number): GoalMetrics {
  const monthlyToward = Math.max(120, Math.round(net * (g.ratio || 0.5)));
  const remaining = Math.max(0, g.amount - g.saved);
  const monthsNeeded = monthlyToward > 0 ? Math.ceil(remaining / monthlyToward) : 99;
  const ahead = Math.round(g.month) - monthsNeeded;
  const tone = ahead >= 0 ? '#2F7D5B' : '#C0792E';
  const pct = Math.max(0, Math.min(100, Math.round((g.saved / g.amount) * 100)));
  return {
    monthlyToward,
    remaining,
    monthsNeeded,
    ahead,
    tone,
    pct,
    pillBg: ahead >= 0 ? 'rgba(47,125,91,0.1)' : 'rgba(192,121,46,0.12)',
    statusLabel: ahead >= 0 ? (ahead === 0 ? 'Right on time' : ahead + ' months early') : Math.abs(ahead) + ' months behind',
    hitDate: monthLabel(monthsNeeded),
    sub:
      ahead >= 0
        ? 'At ' +
          money(monthlyToward) +
          '/mo set aside, you clear the ' +
          money(g.amount) +
          ' target ' +
          (ahead === 0 ? 'right on schedule' : 'ahead of your ' + monthLabel(g.month) + ' date') +
          '.'
        : 'At ' + money(monthlyToward) + '/mo you fall short of your ' + monthLabel(g.month) + ' date. Trim spending or move the date out.',
  };
}
